clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java

```diff
@@ -370,6 +378,200 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
+                        if (buffer.remaining() < 4 && bytesRemaining > 0) {
+                            needMore = true;
+                        } else {
+                            valueSize = ByteUtils.readVarint(buffer);
+                            bytesRead += ByteUtils.sizeOfVarint(valueSize);
+                            if (valueSize > 0) {
+                                if (remainingHeadersToSkip < 0) {
+                                    recordValueSize = valueSize;
+                                }
+                                state = ReadState.SKIP_VALUE;
+                                remainingValueBytesToSkip = valueSize;
+                            } else {
+                                if (remainingHeadersToSkip > 0) {
+                                    remainingHeadersToSkip -= 1;
+                                    if (remainingHeadersToSkip == 0)
+                                        notDone = false;
+                                    else
+                                        state = ReadState.READ_KEY_SIZE;
+                                } else {
+                                    state = ReadState.READ_NUM_HEADERS;
+                                }
+                            }
+                        }
+                        break;
+
+                    case SKIP_VALUE:
+                        // skip bytes until value is consumed
+                        if (remainingValueBytesToSkip > buffer.remaining()) {
+                            remainingValueBytesToSkip -= buffer.remaining();
+                            buffer.position(buffer.limit());
+                            needMore = true;
+                        } else {
+                            buffer.position(buffer.position() + remainingValueBytesToSkip);
+                            bytesRead += valueSize;
+
+                            if (remainingHeadersToSkip > 0) {
+                                remainingHeadersToSkip -= 1;
+                                if (remainingHeadersToSkip == 0)
+                                    notDone = false;
+                                else
+                                    state = ReadState.READ_KEY_SIZE;
+                            } else {
+                                state = ReadState.READ_NUM_HEADERS;
+                            }
+                        }
+                        break;
+
+                    case READ_NUM_HEADERS:
+                        // read num headers (varint, max 4bytes)
+                        if (buffer.remaining() < 4 && bytesRemaining > 0) {
+                            needMore = true;
+                        } else {
+                            int numHeaders = ByteUtils.readVarint(buffer);
+                            bytesRead += ByteUtils.sizeOfVarint(numHeaders);
+                            if (numHeaders > 0) {
+                                state = ReadState.READ_KEY_SIZE;
```

Comment:
I was playing around with some ideas to simplify this parsing logic since we're mostly doing the same thing over and over. I think some helpers might do the trick. This is what I'm thinking:
```java
    private static int readMore(ByteBuffer buffer,
                                DataInput input,
                                int bytesRemaining) throws IOException {
        byte[] array = buffer.array();

        // first copy the remaining bytes to the beginning of the array
        // do not use System.arrayCopy since it will allocate a new array for same src/dest
        int stepsToLeftShift = buffer.position();
        int bytesToLeftShift = buffer.remaining();
        for (int i = 0; i < bytesToLeftShift; i++) {
            array[i] = array[i + stepsToLeftShift];
        }

        // then try to read more bytes to the remaining of the array
        int bytesRead = Math.min(bytesRemaining, array.length - bytesToLeftShift);
        input.readFully(array, bytesToLeftShift, bytesRead);
        buffer.rewind();
        // only those many bytes are readable
        buffer.limit(bytesToLeftShift + bytesRead);
        return bytesRead;
    }

    private static int skipLengthDelimitedField(ByteBuffer buffer,
                                                DataInput input,
                                                int bytesRemaining) throws IOException {
        boolean needMore = false;
        int bytesRead = 0;
        int bytesToSkip = -1;

        while (true) {
            if (needMore) {
                if (bytesRemaining > 0) {
                    int bytesToRead = readMore(buffer, input, bytesRemaining);
                    bytesRemaining -= bytesToRead;
                    needMore = false;
                } else {
                    throw new InvalidRecordException("Invalid record size: expected to read more bytes in record payload");
                }
            }

            if (bytesToSkip < 0) {
                if (buffer.remaining() < 5 && bytesRemaining > 0) {
                    needMore = true;
                } else {
                    bytesToSkip = ByteUtils.readVarint(buffer);
                    if (bytesToSkip <= 0)
                        return bytesRead;
                }
            } else {
                if (bytesToSkip > buffer.remaining()) {
                    bytesToSkip -= buffer.remaining();
                    buffer.position(buffer.limit());
                    needMore = true;
                } else {
                    buffer.position(buffer.position() + bytesToSkip;
                    return bytesRead;
                }
            }
        }
    }
```

With these helpers, the state machine becomes a little simpler. For example:

```java
case READ_KEY:
  bytesRemaining -= skipLengthDelimitedField(buffer, input, bytesRemaining);
  state = ReadState.READ_VALUE;
case READ_VALUE:
  bytesRemaining -= skipLengthDelimitedField(buffer, input, bytesRemaining);
  state = ReadState.READ_NUM_HEADERS;
...
```
If we could get this working, then we may not need to reuse states.